### DISCUSSION TITLE
Lost getting SOC from Mitsubishi iMiev

### DISCUSSION THREAD

**Discussion by raphael303** [meatpiHQ/wican-fw#65] [Q&A]
Hi
So I got MeatPi connected and I am sending requests from Node Red.

Right now my request looks like this:
{ "bus": "0", "type": "tx", "frame": [{ "id": 1889, "dlc": 8, "rtr": false, "extd": false, "data": [2, 21, 1, 170, 170, 170, 170, 170] }] };

The answer I get (like a 100 times) is:
{"bus":"0","type":"rx","ts":26171,"frame":[{"id":1060,"dlc":8,"rtr":false,"extd":false,"data":[64,0,128,0,133,221,1,255]}]}

I have derived the ID, Mode and PID from here:
https://github.com/puddnig/triplet-bmu/blob/master/triplet.py
Which is a python script which reads the SOC and more from the Mitsubishi iMiev

The ID I have derived from, the command 'atsh761' I believe it sets the header to 761, which is the ID for the BMU in hexadecimal. In decimal, this would be 1889.

Then the python script does this:
```
        elif command == '022101':
            self.bmu_data['SOC'] = int(content[0:2], 16)/2.0-5
            self.bmu_data['SOC Display'] = int(content[2:4], 16)/2.0-5
            self.bmu_data['Cell Voltage max'] = voltage(4, 8)
            self.bmu_data['Cell Voltage max ID'] = int(content[8:10], 16) + 1
            self.bmu_data['Cell Voltage min'] = voltage(10, 14)
            self.bmu_data['Cell Voltage min ID'] = int(content[14:16], 16) + 1
            self.bmu_data['Battery Voltage'] = int(content[16:20], 16)/10.0
            self.bmu_data['Temp max'] = temp(20, 22)
            self.bmu_data['Temp max module ID'] = int(content[22:24], 16) + 1
            self.bmu_data['Temp min'] = temp(24, 26)
            self.bmu_data['Temp min module ID'] = int(content[26:28], 16) + 1
            self.bmu_data['Battery Capacity full'] = hex2int(54, 58)/10.0
            self.bmu_data['Battery Capacity currently'] = hex2int(58, 62)/10.0
            self.bmu_data['Battery max input power'] = hex2int(62, 64)/4.0
            self.bmu_data['Battery max output power'] = hex2int(64, 66)/4.0
            self.bmu_data['Target voltage balancing'] = voltage(70, 74)
            self.bmu_data['BMU control voltage'] = hex2int(74, 76)/10.0
            self.bmu_data['EV-ECU control voltage'] = hex2int(76, 78)/10.0
            self.bmu_data['Battery fan pwm %'] = hex2int(78, 80)

```
I believe The command '022101' can be broken down as follows:

02: Number of additional bytes that follow
21: Mode (Service 21 is for proprietary diagnostics)
01: PID (Parameter ID)

Hence my request:
{ "bus": "0", "type": "tx", "frame": [{ "id": 1889, "dlc": 8, "rtr": false, "extd": false, "data": [2, 21, 1, 170, 170, 170, 170, 170] }] };
and response:
{"bus":"0","type":"rx","ts":26171,"frame":[{"id":1060,"dlc":8,"rtr":false,"extd":false,"data":[64,0,128,0,133,221,1,255]}]}

From other ressources on the web I have found that the SOC is PID x374. I have thus tried (decimal):
msg.payload = { "bus": "0", "type": "tx", "frame": [{ "id": 1889, "dlc": 8, "rtr": false, "extd": false, "data": [2, 21, 884, 170, 170, 170, 170, 170] }] };
response:
{"bus":"0","type":"rx","ts":21588,"frame":[{"id":1060,"dlc":8,"rtr":false,"extd":false,"data":[64,0,128,0,133,239,1,255]}]}

I have converted the decimal strings to hex and tried to apply the logic of the above python script, but the results are nonsense.

Any help is appreciated.

**Comment by raphael303** [comment]
In case anyone ever does it:
MeatPi wiCAN interface, sends CAN Data via MQTT to home Assistant. Node Red Subscribed to topic.
Function with 16 Outputs sends iMiev Data to Home Assistant for automation.
Note, the IDs are in decimal.

Here is the code for the function:
```

var SOC1, SOC2, Battery100Cap, CellMaxTemp, CellMinTemp, BatteryVolts, CellMaxVolt, CellMinVolt, ChargeAmpsAC, ChargeAmpsDC, ChargeVoltsAC, ChargeVoltsDC, ChargerTemperature1, ChargerTemperature2, Odometer, OutdoorAirTemperature;

for (let i = 0; i < msg.payload.frame.length; i++) {
    let frame = msg.payload.frame[i];
    let id = frame.id;
    let data = frame.data;

    switch (id) {
        case 884:
            SOC1 = { payload: (data[0] - 10) / 2 };
            SOC2 = { payload: (data[1] - 10) / 2 };
            Battery100Cap = { payload: data[6] / 2 };
            CellMaxTemp = { payload: data[4] - 50 };
            CellMinTemp = { payload: data[5] - 50 };
            break;


        case 883:
            BatteryVolts = { payload: (data[4] * 256 + data[5]) / 10 };
            CellMaxVolt = { payload: (data[0] + 210) / 100 };
            CellMinVolt = { payload: (data[1] + 210) / 100 };
            break;
        case 905:
            ChargeAmpsAC = { payload: data[6] / 10 };
            ChargeAmpsDC = { payload: data[2] / 10 };
            ChargeVoltsAC = { payload: data[1] };
            ChargeVoltsDC = { payload: 2 * data[0] + 0.5 };
            ChargerTemperature1 = { payload: data[3] - 50 };
            ChargerTemperature2 = { payload: data[4] - 50 };
            break;
        case 1042:
            Odometer = { payload: data[2] * 65536 + data[3] * 256 + data[4] };
            break;
        case 646:
            OutdoorAirTemperature = { payload: data[3] - 50 };
            break;
        default:
            break;
    }
}

return [SOC1, SOC2, Battery100Cap, CellMaxTemp, CellMinTemp, BatteryVolts, CellMaxVolt, CellMinVolt, ChargeAmpsAC, ChargeAmpsDC, ChargeVoltsAC, ChargeVoltsDC, ChargerTemperature1, ChargerTemperature2, Odometer, OutdoorAirTemperature];


```
However, you only get data, when it is charging. I did not find out, yet, how to request that data while idle without ignition. I have connection and get data, just not that data.

**Reply by raphael303** [reply]
However, this is of no use, because the sheer amount of data almost crashes nodered on HA on rpi4. One wican might work more or less. Two is too much.

**Reply by meatpiHQ** [reply]
@raphael303 thanks for sharing. I'll have a solution for #66 soon.
